File: shared/src/main/java/ru/privatenull/pnauth/security/IpBanStore.java
package ru.privatenull.pnauth.security;

import java.time.Clock;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class IpBanStore {
    private final Map<String, Long> bans = new ConcurrentHashMap<>();
    private final Clock clock;

    public IpBanStore() {
        this(Clock.systemUTC());
    }

    public IpBanStore(Clock clock) {
        this.clock = clock;
    }

    public void ban(String ip, Duration duration) {
        if (ip != null && !ip.isBlank()) bans.put(ip, clock.millis() + duration.toMillis());
    }

    public boolean isBanned(String ip) {
        Long until = bans.get(ip);
        if (until == null) return false;
        if (until <= clock.millis()) {
            bans.remove(ip, until);
            return false;
        }
        return true;
    }

    public void clear() {
        bans.clear();
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/security/IpBanStore.kt
package ru.privatenull.pnauth.security

import java.time.Clock
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap

class IpBanStore @JvmOverloads constructor(
    private val clock: Clock = Clock.systemUTC()
) {
    private val bans: MutableMap<String, Long> = ConcurrentHashMap()

    fun ban(ip: String?, duration: Duration) {
        if (!ip.isNullOrBlank()) {
            bans[ip] = clock.millis() + duration.toMillis()
        }
    }

    fun isBanned(ip: String?): Boolean {
        if (ip == null) return false
        val until = bans[ip] ?: return false
        if (until <= clock.millis()) {
            bans.remove(ip, until)
            return false
        }
        return true
    }

    fun clear() {
        bans.clear()
    }
}
